processor/src/main/java/org/mapstruct/ap/internal/model/source/TargetReference.java

```
@@ -142,46 +173,150 @@ public TargetReference build() {
                     // the last may be other readAccessor (setter / getter / adder).
 
-                    if ( Executables.isGetterMethod( targetWriteAccessor ) ||
-                        Executables.isFieldAccessor( targetWriteAccessor ) ) {
-                        nextType = typeFactory.getReturnType(
-                            (DeclaredType) nextType.getTypeMirror(),
-                            targetWriteAccessor );
-                    }
-                    else {
-                        nextType = typeFactory.getSingleParameter(
-                            (DeclaredType) nextType.getTypeMirror(),
-                            targetWriteAccessor ).getType();
-                    }
+                    nextType = findNextType( nextType, targetWriteAccessor, targetReadAccessor, isLast );
 
                     // check if an entry alread exists, otherwise create
                     String[] fullName = Arrays.copyOfRange( entryNames, 0, i + 1 );
                     PropertyEntry propertyEntry = PropertyEntry.forTargetReference( fullName, targetReadAccessor,
                         targetWriteAccessor, nextType );
                     targetEntries.add( propertyEntry );
-                    }
-
                 }
 
+            }
+
             return targetEntries;
         }
 
+        /**
+         * Finds the next type based on the initial type.
+         *
+         * @param initial for which a next type should be found
+         * @param targetWriteAccessor the write accessor
+         * @param targetReadAccessor the read accessor
+         * @param isLast whether this is the call for the last accessor
+         *
+         * @return the next type that should be used for finding a property entry
+         */
+        private Type findNextType(Type initial, Accessor targetWriteAccessor,
+            Accessor targetReadAccessor, boolean isLast) {
+            if ( !isLast && targetWriteAccessor == null ) {
```

I would remove it. If you made a mistake. its sometimes better to let it fail drastically e.g. by a null pointer.